node_modules/oae-principals/lib/api.user.js

```diff
@@ -119,63 +126,275 @@ var createUser = module.exports.createUser = function(ctx, tenantAlias, displayN
+ * @param  {Function}       callback                    Standard callback function
+ * @param  {Object}         callback.err                An error that occurred, if any
+ * @param  {Object}         callback.principalIds       Object mapping each principal of `principals` to a `principalId`
+ * @api private
+ */
+var _createGuestUsers = function(ctx, principals, callback, _principalIds) {
+    _principalIds = _principalIds || {};
+    if (principals.length === 0) {
+        return callback(null, _principalIds);
+    }
+
+    var principal = principals.pop();
+    // If the `principal` is a user or group id, we can move on to the next one
+    if (AuthzUtil.isUserId(principal) || AuthzUtil.isGroupId(principal)) {
+        _principalIds[principal] = principal;
+        return _createGuestUsers(ctx, principals, callback, _principalIds);
+    }
+
+    // The `principal` is an email
+    createGuestUser(ctx, principal, function(err, user) {
         if (err) {
             return callback(err);
         }
 
-        var createdUser = new PrincipalsModel.User(tenantAlias, id, displayName, opts);
-        createdUser.needsToAcceptTC = false;
+        _principalIds[principal] = user.id;
+        _createGuestUsers(ctx, principals, callback, _principalIds);
+    });
+};
+
+/**
+ * Verify a guest token and associate a local login strategy with the user matching the token.
+ *
+ * @param  {Context}    ctx                         Standard context object containing the current user and the current tenant
+ * @param  {String}     token                       The token to verify
+ * @param  {String}     username                    The desired username
+ * @param  {String}     password                    The desired password
+ * @param  {Function}   callback                    Standard callback function
+ * @param  {Object}     callback.err                An error that occurred, if any
+ */
+var verifyGuestToken = module.exports.verifyGuestToken = function(ctx, token, username, password, callback) {
+    var validator = new Validator();
+    validator.check(token, {'code': 400, 'msg': 'A token must be provided'}).notEmpty();
+    validator.check(username, {'code': 400, 'msg': 'A username must be provided'}).notEmpty();
+    validator.check(password, {'code': 400, 'msg': 'A password must be provided'}).notEmpty();
+    validator.check(password, {'code': 400, 'msg': 'A password must contain at least 6 characters'}).len(6);
+    if (validator.hasErrors()) {
+        return callback(validator.getFirstError());
+    }
 
-        PrincipalsEmitter.emit(PrincipalsConstants.events.CREATED_USER, ctx, createdUser);
-        return callback(null, createdUser);
+    PrincipalsDAO.getUserViaEmailToken(token, function(err, user) {
+        if (err) {
+            return callback(err);
+        }
+
+        // Associate the user with the new login
+        var hashedPassword = require('oae-authentication/lib/util').hashPassword(password);
```

Comment:
Maybe, I'd rather not do it in this PR though, that would explode the line count :(